[wasm] Build static components; include hot_reload in runtime

lambdageek opened the pull request:
Workaround until https://github.com/dotnet/runtime/issues/54565 is fixed
Build the runtime always with support for hot_reload, and without diagnostics_tracing

lambdageek committed:
[wasm] Build static components; include hot_reload in runtime

Workaround until https://github.com/dotnet/runtime/issues/54565 is fixed
Build the runtime always with support for hot_reload, and without diagnostics_tracing

lambdageek commented:
/cc @lateralusX

lambdageek commented on src/mono/wasm/Makefile:
```diff
@@ -48,12 +48,19 @@ provision-wasm: .stamp-wasm-install-and-select-$(EMSCRIPTEN_VERSION)
 	$(MONO_BIN_DIR)/libmono-ee-interp.a \
 	$(MONO_BIN_DIR)/libmonosgen-2.0.a \
+	$(MONO_COMPONENT_LIBS) \
+	$(MONO_BIN_DIR)/libmonosgen-2.0.a \
```
libmonosgen-2.0 is here twice intentionally - the components depend on the runtime and the runtime depends on the components.

radical replied:
add a comment mentioning this?

lambdageek commented:
This also needs to go into .NET 6 Preview 6

radical commented:
Does this enable, or fix something?

lambdageek commented:
> Does this enable, or fix something?

Yea it fixes hot reload on .NET 6 P6

radical commented:
> > Does this enable, or fix something?
> 
> Yea it fixes hot reload on .NET 6 P6

Is there a test that we can enable with this, so we get validation too?

thaystg committed:
Update wasm.proj

lambdageek committed:
Add a browser functional test for hot reload

Just check that the capabilities are non-empty which is a good proxy for hot
reload being enabled in the runtime.

lambdageek commented:
@radical I added a new functional test for the browser.  didn't try very hard to override all the build lane stuff, but at least for a default `./build.sh --os browser -c Release` and running the test with `dotnet.sh build /p:TargetOS=browser /p:TargetArchitecture=wasm /t:Test` works for me.

lambdageek committed:
Turn off trimming for hot reload functional test

lambdageek committed:
Why is this necessary? This assembly shouldn't be trimmed

lambdageek committed:
No really, don't trim the modifiable assemblies

radical commented:
wasm failure is https://github.com/dotnet/runtime/issues/53957

lambdageek committed:
Disable test on browser AOT

radical commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdateTest.cs:
```diff
@@ -18,6 +18,7 @@ namespace System.Reflection.Metadata
     public class ApplyUpdateTest
     {
         [Fact]
+	[ActiveIssue("https://github.com/dotnet/runtime/issues/54617", typeof(PlatformDetection), nameof(PlatformDetection.IsBrowser), nameof(PlatformDetection.IsMonoAOT))]
```
nit: spaces instead of tab

radical commented on src/mono/wasm/wasm.proj:
```diff
@@ -189,13 +189,16 @@
                              $(NativeBinDir)libSystem.Native.a;
                              $(NativeBinDir)libSystem.IO.Compression.Native.a" />
       <MonoLibFiles Include="@(ICULibNativeFiles)" />
+      <MonoLibFiles Include="@(MonoComponentLibs)" />        
```
```suggestion
      <MonoLibFiles Include="@(MonoComponentLibs)" />
```

lambdageek committed:
fix whitespace

radical commented on src/tests/FunctionalTests/WebAssembly/Browser/HotReload/ApplyUpdateReferencedAssembly/ApplyUpdateReferencedAssembly.csproj:
```diff
@@ -0,0 +1,30 @@
+
+  <ItemGroup>
+    <!-- This package from https://github.com/dotnet/hotreload-utils provides
+	 targets that read the json delta script and generates deltas based on the baseline assembly and the modified sources.
+
+	 Projects must define the DeltaScript property that specifies the (relative) path to the json script.
+	 Deltas will be emitted next to the output assembly.  Deltas will be copied when the current
+	 project is referenced from other other projects.
```
```suggestion
     targets that read the json delta script and generates deltas based on the baseline assembly and the modified sources.

     Projects must define the DeltaScript property that specifies the (relative) path to the json script.
     Deltas will be emitted next to the output assembly.  Deltas will be copied when the current
     project is referenced from other other projects.
```

radical commented on src/tests/FunctionalTests/WebAssembly/Browser/HotReload/WebAssembly.Browser.HotReload.Test.csproj:
```diff
@@ -0,0 +1,52 @@
+	  Condition="'$(TargetOS)' == 'Browser'">
+    <ItemGroup>
+      <!-- FIXME: this belongs in eng/testing/tests.wasm.targets -->
+      <!-- FIXME: Can we do something on the Content items in the referenced projects themselves to get this for free? -->
```
soon ;)

radical commented on src/tests/FunctionalTests/WebAssembly/Browser/HotReload/WebAssembly.Browser.HotReload.Test.csproj:
```diff
@@ -0,0 +1,52 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
tabs -> spaces for this whole file

radical commented on src/tests/FunctionalTests/WebAssembly/Browser/HotReload/Program.cs:
```diff
@@ -0,0 +1,82 @@
+            return success;
+        }
+
+
```
```suggestion
```

radical reviewed (approved):
Looks good, just nits, whatever bits I could review. The env vars stuff should be doable in the project file itself, but nothing to block this PR over.

lambdageek committed:
tabs->spaces

lambdageek commented:
/backport to release/6.0-preview6
